yoyodyne/datasets.py

```diff
@@ -229,6 +229,43 @@ def decode_target(
     def has_features(self) -> bool:
         return self.index.has_features
 
+    @staticmethod
+    def read_index(path: str) -> indexes.Index:
+        """Helper for loading index.
+
+        Args:
+            path (str).
+
+        Returns:
+            indexes.IndexNoFeatures.
+        """
+        return indexes.Index.read(path)
+
+    def __getitem__(self, idx: int) -> Item:
```

Comment:
I think you need to reimplement this in the `DatasetFeatures` class too, since this does not return encoded features.

Reply:
https://github.com/CUNY-CL/yoyodyne/blob/25212bc06f848f8c8b18d3d185655e553725f326/yoyodyne/datasets.py <- this is the same setup as premerge, unless the issue was introduced there?

Reply:
Isn't this in the `DatasetNoFeatures` class? What is the expected behavior?

Reply:
This is in the `DataSetNoFeatures`. I thought you'd need the the impl [here](https://github.com/CUNY-CL/yoyodyne/blob/25212bc06f848f8c8b18d3d185655e553725f326/yoyodyne/datasets.py#L287) for `DatasetFeatures` otherwise it'll inherit the no-features one (which is clearly insufficient).

If you think you don't need to, install from your PR and test the pointer generator on a language with features (Polish abstractness data?) to confirm.

Reply:
Sorry I misread your original comment. `DatasetFeatures` does not inherit `__getitem__` in master https://github.com/CUNY-CL/yoyodyne/blob/master/yoyodyne/datasets.py#L262

So it looks like the issue is that the method was just deleted from `DatasetNoFeatures` if I am not mistaken? Indeed we appear to have both in this branch too: https://github.com/bonham79/yoyodyne/blob/feature_update_hotfix/yoyodyne/datasets.py

Reply:
it is re-implemented, see line 299.  It's just not showing up in the merge comparison.

Reply:
```
Using bfloat16 Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/home/salamander/anaconda3/envs/yoyodyne/lib/python3.10/site-packages/pytorch_lightning/trainer/setup.py:176: PossibleUserWarning: GPU available but not used. Set `accelerator` and `devices` using `Trainer(accelerator='gpu', devices=2)`.
  rank_zero_warn(
Including features
Source vocabulary: '<UNK>', '<P>', '<S>', '<E>', "'", '-', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'w', 'x', 'y', 'z', 'ó', 'ą', 'ć', 'ę', 'ł', 'ń', 'ś', 'ź', 'ż'
Feature vocabulary: '<UNK>', '<P>', '<S>', '<E>', '[ANIM]', '[FEM]', '[HUM]', '[INAN]', '[MASC]', '[NEUT]'
Target vocabulary: '<UNK>', '<P>', '<S>', '<E>', "'", '-', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'w', 'x', 'y', 'z', 'ó', 'ą', 'ć', 'ę', 'ł', 'ń', 'ś', 'ź', 'ż'
Index: pol/pol/index.pkl
Model: PointerGeneratorLSTMEncoderDecoder
Model: LSTMEncoder

```

Reply:
> it is re-implemented, see line 299. It's just not showing up in the merge comparison.

Yes, sorry, that is what I am saying :).